File: tron-backend/src/adapters/ai.js
const { OpenAI } = require('openai');

// Initialize the client pointing to OpenRouter instead of OpenAI!
const openai = new OpenAI({
    baseURL: "https://openrouter.ai/api/v1",
    apiKey: process.env.OPENROUTER_API_KEY,
    defaultHeaders: {
        "HTTP-Referer": "http://localhost:3000", // Required by OpenRouter
        "X-Title": "T.R.O.N. Local Watcher", // Required by OpenRouter
    }
});

// ==========================================
// THE INTELLIGENCE ENGINE
// ==========================================
async function generateExecutiveSummary(prTitle, sanitizedDiff) {
    // 🛡️ QoL UPDATE: Don't waste AI credits on empty diffs
    if (!sanitizedDiff || sanitizedDiff.trim().length === 0) {
        console.log(`\n⏭️  [AI ADAPTER] Diff is empty after sanitization. Skipping LLM.`);
        return {
            intent: "Infrastructure",
            executive_summary: "Automated updates to lockfiles, generated assets, or ignored files.",
            business_impact: "No direct user impact. Routine maintenance.",
            confidence_score: 100
        };
    }
    console.log(`\n🤖 [AI ADAPTER] Analyzing code diff via OpenRouter for PR: "${prTitle}"...`);

    const systemPrompt = `
    You are an elite Staff Software Engineer translating technical code changes into business intelligence.
    Read the provided Git diff and summarize exactly what changed and why it matters.
    
    RULES:
    1. Do not use overly technical jargon.
    2. Focus on the business value.
    3. You MUST respond in pure, raw JSON format matching the exact structure below.
    
    JSON STRUCTURE:
    {
        "intent": "Feature" | "Bug Fix" | "Refactoring" | "Infrastructure",
        "executive_summary": "A 2-3 sentence human-readable summary of the changes.",
        "business_impact": "A 1 sentence explanation of how this affects the user.",
        "confidence_score": <number between 1 and 100 representing how certain you are of this summary>
    }
    `;

    try {
        const response = await openai.chat.completions.create({
            model: "openrouter/free", 
            messages: [
                { role: "system", content: systemPrompt },
                { role: "user", content: `PR Title: ${prTitle}\n\nCode Diff:\n${sanitizedDiff}` }
            ],
            temperature: 0.1 
        });

        let rawContent = response.choices[0].message.content.trim();
        const jsonMatch = rawContent.match(/\{[\s\S]*\}/);
        if (!jsonMatch) {
            throw new Error("AI did not return a valid JSON structure.");
        }

        const aiResult = JSON.parse(jsonMatch[0]);
        
        console.log(`✅ [AI ADAPTER] Analysis Complete! Intent: ${aiResult.intent}`);
        return aiResult;

    } catch (error) {
        console.error(`❌ [AI ADAPTER] Failed to generate summary:`, error.message);
        return {
            intent: "Unknown",
            executive_summary: `Developer opened PR: ${prTitle}. AI analysis failed or timed out.`,
            business_impact: "Requires manual review.",
            confidence_score: 0 
        };
    }
}

// ==========================================
// 🕵️‍♂️ NEW: THE CODE REVIEWER ENGINE
// ==========================================
async function generateCodeReview(sanitizedDiff) {
    if (!sanitizedDiff || sanitizedDiff.trim().length === 0) return "✅ **No code changes detected.**";

    console.log(`🤖 [AI ADAPTER] Performing deep technical code review...`);

    const reviewPrompt = `
    You are a strict, senior software engineer performing a technical code review.
    Review the following Git diff for logic errors, security vulnerabilities, or performance bottlenecks.
    
    RULES:
    1. If the code is perfect, respond ONLY with: "✅ **Code looks solid.** No major issues detected by T.R.O.N."
    2. If there are issues, list them using Markdown bullet points. 
    3. Be direct, technical, and helpful. Use code blocks for suggestions.
    4. Limit your response to the top 3 most critical findings.
    `;

    try {
        const response = await openai.chat.completions.create({
            model: "openrouter/free", 
            messages: [
                { role: "system", content: reviewPrompt },
                { role: "user", content: `Code Diff:\n${sanitizedDiff}` }
            ],
            temperature: 0.3 // Slightly higher for more creative debugging insight
        });

        return response.choices[0].message.content.trim();

    } catch (error) {
        console.error(`❌ [AI ADAPTER] Code Review Failed:`, error.message);
        return "⚠️ T.R.O.N. was unable to generate a code review due to a technical error.";
    }
}

// ==========================================
// ✨ NEW: THE TASK SUGGESTION ENGINE
// ==========================================
async function generateTaskSuggestions(diffData) {
    if (!diffData || diffData.trim().length === 0) return [];

    console.log(`🤖 [AI ADAPTER] Brainstorming tasks for uncommitted code...`);

    const prompt = `
    You are a Senior Technical Product Manager. A developer is currently writing some code, but they haven't created a ticket for it yet.
    Look at the following raw, uncommitted code diff and suggest 3 short, actionable task titles (under 8 words each) that accurately describe what the developer is building.
    
    RULES:
    1. Start each task with a strong action verb (e.g., Implement, Fix, Add, Refactor).
    2. Keep them short and professional (like Basecamp or Jira ticket titles).
    3. You MUST respond in pure JSON format as an array of exactly 3 strings. Do not add markdown or extra text.
    
    EXAMPLE OUTPUT:
    ["Implement user authentication middleware", "Fix database connection retry logic", "Add error boundaries to React router"]

    Code Diff:
    ${diffData}
    `;

    try {
        const response = await openai.chat.completions.create({
            model: "openrouter/free", 
            messages: [{ role: "user", content: prompt }],
            temperature: 0.4 // Just enough creativity to guess the intent
        });

        let rawContent = response.choices[0].message.content.trim();
        const jsonMatch = rawContent.match(/\[[\s\S]*\]/);
        if (!jsonMatch) throw new Error("AI did not return a valid JSON array.");

        const suggestions = JSON.parse(jsonMatch[0]);
        return suggestions.slice(0, 3); // Guarantee we only return 3

    } catch (error) {
        console.error(`❌ [AI ADAPTER] Task Suggestion Failed:`, error.message);
        // Fallback so the VS Code UI doesn't crash
        return ["Implement new feature", "Refactor codebase", "Fix code issue"]; 
    }
}

// ==========================================
// 🛡️ THE WRAPPER FUNCTION FOR TEST SCRIPT & DISCORD
// ==========================================
async function generateSummary(diffData) {
    const aiJsonResult = await generateExecutiveSummary("Manual Diff Test", diffData);
    
    return `
**🎯 Intent:** ${aiJsonResult.intent}
**💼 Business Impact:** ${aiJsonResult.business_impact}
**📝 Summary:** ${aiJsonResult.executive_summary}
**🛡️ Confidence Score:** ${aiJsonResult.confidence_score || 0}/100
    `.trim();
}

// 🛡️ Export all functions
module.exports = { 
    generateExecutiveSummary, 
    generateCodeReview,
    generateSummary,
    generateTaskSuggestions
};

File: tron-go-backend/internal/adapters/ai.go
package adapters

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"time"
)

// AIAdapter handles communication with OpenRouter
type AIAdapter struct {
	HTTPClient *http.Client
	APIKey     string
	BaseURL    string
	Model      string
}

// ExecutiveSummary matches the exact JSON schema requested by the prompt
type ExecutiveSummary struct {
	Intent           string `json:"intent"`
	ExecutiveSummary string `json:"executive_summary"`
	BusinessImpact   string `json:"business_impact"`
	ConfidenceScore  int    `json:"confidence_score"`
}

// NewAIAdapter initializes the client for OpenRouter
func NewAIAdapter() *AIAdapter {
	apiKey := os.Getenv("OPENROUTER_API_KEY")
	if apiKey == "" {
		fmt.Println("⚠️  Warning: OPENROUTER_API_KEY is not set in environment")
	}

	return &AIAdapter{
		HTTPClient: &http.Client{Timeout: 45 * time.Second}, // 45s timeout for AI generation
		APIKey:     apiKey,
		BaseURL:    "https://openrouter.ai/api/v1/chat/completions",
		Model:      "openrouter/free", // 🌟 Using your specified model
	}
}

// --------------------------------------------------------
// 🛠️ PRIVATE HELPER: Executes the OpenRouter API Call
// --------------------------------------------------------
func (api *AIAdapter) executeChat(messages []map[string]string, temperature float64) (string, error) {
	if api.APIKey == "" {
		return "", fmt.Errorf("OPENROUTER_API_KEY is missing")
	}

	payload := map[string]interface{}{
		"model":       api.Model,
		"messages":    messages,
		"temperature": temperature,
	}
	payloadBytes, _ := json.Marshal(payload)

	req, err := http.NewRequest("POST", api.BaseURL, bytes.NewBuffer(payloadBytes))
	if err != nil {
		return "", err
	}

	// 🌟 OpenRouter Required Headers
	req.Header.Set("Authorization", "Bearer "+api.APIKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("HTTP-Referer", "http://localhost:3000")
	req.Header.Set("X-Title", "T.R.O.N. Local Watcher")

	resp, err := api.HTTPClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("AI network error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("OpenRouter API error (%d): %s", resp.StatusCode, string(bodyBytes))
	}

	var result struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("failed to decode AI response: %w", err)
	}

	if len(result.Choices) == 0 {
		return "", fmt.Errorf("OpenRouter returned an empty response")
	}

	return result.Choices[0].Message.Content, nil
}

// --------------------------------------------------------
// 1. THE INTELLIGENCE ENGINE (JSON)
// --------------------------------------------------------
func (api *AIAdapter) GenerateExecutiveSummary(prTitle, sanitizedDiff string) (ExecutiveSummary, error) {
	// QoL UPDATE: Don't waste AI credits on empty diffs
	if sanitizedDiff == "" {
		return ExecutiveSummary{
			Intent:           "Infrastructure",
			ExecutiveSummary: "Automated updates to lockfiles, generated assets, or ignored files.",
			BusinessImpact:   "No direct user impact. Routine maintenance.",
			ConfidenceScore:  100,
		}, nil
	}

	systemPrompt := `You are an elite Staff Software Engineer translating technical code changes into business intelligence.
Read the provided Git diff and summarize exactly what changed and why it matters.

RULES:
1. Do not use overly technical jargon.
2. Focus on the business value.
3. You MUST respond in pure, raw JSON format matching the exact structure below.

JSON STRUCTURE:
{
	"intent": "Feature" | "Bug Fix" | "Refactoring" | "Infrastructure",
	"executive_summary": "A 2-3 sentence human-readable summary of the changes.",
	"business_impact": "A 1 sentence explanation of how this affects the user.",
	"confidence_score": 95
}`

	userPrompt := fmt.Sprintf("PR Title: %s\n\nCode Diff:\n%s", prTitle, sanitizedDiff)

	messages := []map[string]string{
		{"role": "system", "content": systemPrompt},
		{"role": "user", "content": userPrompt},
	}

	rawContent, err := api.executeChat(messages, 0.1)
	if err != nil {
		return ExecutiveSummary{}, err
	}

	// 🛡️ Regex extraction to handle markdown wrappers (e.g. ```json ... ```)
	re := regexp.MustCompile(`\{[\s\S]*\}`)
	jsonMatch := re.FindString(rawContent)
	if jsonMatch == "" {
		return ExecutiveSummary{}, fmt.Errorf("AI did not return a valid JSON structure")
	}

	var summary ExecutiveSummary
	if err := json.Unmarshal([]byte(jsonMatch), &summary); err != nil {
		return ExecutiveSummary{}, fmt.Errorf("failed to parse AI JSON: %w", err)
	}

	return summary, nil
}

// --------------------------------------------------------
// 2. THE CODE REVIEWER ENGINE (MARKDOWN)
// --------------------------------------------------------
func (api *AIAdapter) GenerateCodeReview(sanitizedDiff string) string {
	if sanitizedDiff == "" {
		return "✅ **No code changes detected.**"
	}

	reviewPrompt := `You are a strict, senior software engineer performing a technical code review.
Review the following Git diff for logic errors, security vulnerabilities, or performance bottlenecks.

RULES:
1. If the code is perfect, respond ONLY with: "✅ **Code looks solid.** No major issues detected by T.R.O.N."
2. If there are issues, list them using Markdown bullet points. 
3. Be direct, technical, and helpful. Use code blocks for suggestions.
4. Limit your response to the top 3 most critical findings.`

	messages := []map[string]string{
		{"role": "system", "content": reviewPrompt},
		{"role": "user", "content": fmt.Sprintf("Code Diff:\n%s", sanitizedDiff)},
	}

	response, err := api.executeChat(messages, 0.3)
	if err != nil {
		fmt.Printf("❌ Code Review Failed: %v\n", err)
		return "⚠️ T.R.O.N. was unable to generate a code review due to a technical error."
	}

	return response
}

// --------------------------------------------------------
// 3. THE TASK SUGGESTION ENGINE (JSON ARRAY)
// --------------------------------------------------------
func (api *AIAdapter) GenerateTaskSuggestions(diffData string) []string {
	fallback := []string{"Implement new feature", "Refactor codebase", "Fix code issue"}

	if diffData == "" {
		return fallback
	}

	prompt := fmt.Sprintf(`You are a Senior Technical Product Manager. A developer is currently writing some code, but they haven't created a ticket for it yet.
Look at the following raw, uncommitted code diff and suggest 3 short, actionable task titles (under 8 words each) that accurately describe what the developer is building.

RULES:
1. Start each task with a strong action verb (e.g., Implement, Fix, Add, Refactor).
2. Keep them short and professional.
3. You MUST respond in pure JSON format as an array of exactly 3 strings. Do not add markdown or extra text.

Code Diff:
%s`, diffData)

	messages := []map[string]string{
		{"role": "user", "content": prompt},
	}

	rawContent, err := api.executeChat(messages, 0.4)
	if err != nil {
		fmt.Printf("❌ Task Suggestion Failed: %v\n", err)
		return fallback
	}

	// 🛡️ Regex to extract the JSON Array
	re := regexp.MustCompile(`\[[\s\S]*\]`)
	jsonMatch := re.FindString(rawContent)
	if jsonMatch == "" {
		return fallback
	}

	var suggestions []string
	if err := json.Unmarshal([]byte(jsonMatch), &suggestions); err != nil {
		return fallback
	}

	// Ensure we return at most 3 elements
	if len(suggestions) > 3 {
		return suggestions[:3]
	}
	return suggestions
}
